src/Components/Web.JS/src/JSInitializers/JSInitializers.ts

```diff
@@ -29,19 +55,97 @@ export class JSInitializer {
       if (initializer === undefined) {
         return;
       }
-      const { beforeStart: beforeStart, afterStarted: afterStarted } = initializer;
-      if (afterStarted) {
-        jsInitializer.afterStartedCallbacks.push(afterStarted);
+
+      if (!jsInitializer.singleRuntime) {
+        return runMultiRuntimeInitializers(jsInitializer, initializer, initializerArguments);
+      } else {
+        const { beforeStart, afterStarted } = initializer;
+        return runClassicInitializers(jsInitializer, beforeStart, afterStarted, initializerArguments);
+      }
+
+      function runMultiRuntimeInitializers(
+        jsInitializer: JSInitializer,
+        initializerModule: Partial<BlazorInitializer>, initializerArguments: unknown[]): void | PromiseLike<void> {
+        const options = initializerArguments[0] as WebStartOptions;
+        const { beforeStart, afterStarted, beforeWebStart, afterWebStarted, beforeWebAssemblyStart, afterWebAssemblyStarted, beforeServerStart, afterServerStarted } = initializerModule;
+        const runtimeSpecificExports = !!(beforeWebStart || afterWebStarted || beforeWebAssemblyStart || afterWebAssemblyStarted || beforeServerStart || afterServerStarted);
+        const hasOnlyClassicInitializers = !!(!runtimeSpecificExports && (beforeStart || afterStarted));
+        const runLegacyInitializers = hasOnlyClassicInitializers && options.enableClassicInitializers;
+        if (hasOnlyClassicInitializers && !options.enableClassicInitializers) {
+          // log warning "classic initializers will be ignored when multiple runtimes are used".
+          // Skipping "adjustedPath" initializer.
+          jsInitializer.logger?.log(
+            LogLevel.Warning,
+            `Initializer '${adjustedPath}' will be ignored because multiple runtimes are available. use 'before(web|webAssembly|server)Start' and 'after(web|webAssembly|server)Started?' instead.)`
+          );
+        } else if (runLegacyInitializers) {
+          return runClassicInitializers(jsInitializer, beforeStart, afterStarted, initializerArguments);
+        }
+
+        ensureInitializers(options);
+
+        if (beforeWebAssemblyStart) {
+          options.webAssembly.initializers.beforeStart.push(beforeWebAssemblyStart);
+        }
+
+        if (afterWebAssemblyStarted) {
+          options.webAssembly.initializers.afterStarted.push(afterWebAssemblyStarted);
+        }
+
+        if (beforeServerStart) {
+          options.circuit.initializers.beforeStart.push(beforeServerStart);
+        }
+
+        if (afterServerStarted) {
+          options.circuit.initializers.afterStarted.push(afterServerStarted);
+        }
+
+        if (afterWebStarted) {
+          jsInitializer.afterStartedCallbacks.push(afterWebStarted);
+        }
+
+        if (beforeWebStart) {
+          return beforeWebStart(options);
+        }
```

Comment:
We essentially collect all the callbacks from the initializer for server and webassembly, and set them in their respective options types. After that, we start running the `beforeWebStart` callback and trigger the `afterWebStarted` callbacks once we are done with the first render.